### ISSUE TITLE
Improve CPU JIT performance.

### ISSUE THREAD

**Issue by mchandra** [arrayfire/arrayfire#1226]
Consider the following code

```
  af::info();

  int N1 = 2048;
  int N2 = 2048;
  int N3 = 1;
  array a = af::randu(N1, N2, N3, f64);
  array b = af::randu(N1, N2, N3, f64);
  array c = af::constant(0., N1, N2, N3, f64);

  int numEvals = 100;
  af::timer::start();
  for (int n=0; n<numEvals; n++)
  {
    c = af::sqrt(a*a + b*a*b);
    c.eval();
    af::sync();
  }
  printf("Time taken for %d automatic evaluations = %g secs\n",
         numEvals, af::timer::stop()
        );

  double *aPtr = a.host<double>();
  double *bPtr = b.host<double>();
  double *cPtr = c.host<double>();

  af::timer::start();
  for (int n=0; n<numEvals; n++)
  {
    for (int j=0; j<N2; j++)
    {   
      for (int i=0; i<N1; i++)
      {   
        cPtr[i + N1*j] = 
          sqrt(  aPtr[i + N1*j] * aPtr[i + N1*j]
               +   bPtr[i + N1*j] * aPtr[i + N1*j] * bPtr[i + N1*j]
              );
      }   
    }   
  }
  printf("Time taken for %d manual evaluations = %g secs\n",
         numEvals, af::timer::stop()
        );
```

Here is the output:

```
#### System info ####
ArrayFire v3.1.3 (CPU, 64-bit Linux, build 35c89f5)
[0] Intel:       Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz Max threads(20) 
Time taken for 100 automatic evaluations = 11.4611 secs
Time taken for 100 manual evaluations = 0.979555 secs
```

Are the numbers as expected?

Thanks,
Mani

**Comment by pavanky** [comment]
Can you time it with the af::sync() outside the loop ?

**Comment by pavanky** [comment]
Nvm, I just checked it on a similar machine. The CPU fall back mechanism is fairly slow (especially the element wise operations) and exists solely for being a fallback mechanism for now. We are investigating ways to improve performance.

**Comment by pavanky** [comment]
With 3.4.1
```
ArrayFire v3.4.2 (CPU, 64-bit Linux, build default)
[0] Intel: Intel(R) Core(TM) i5-6500 CPU @ 3.20GHz, 15985 MB, Max threads(4) 
Time taken for 100 automatic evaluations = 9.9807 secs                                                                                                                                                             Time taken for 100 manual evaluations = 0.970198 secs
```

With 3.5.0
```
ArrayFire v3.5.0 (CPU, 64-bit Linux, build a0dd4f92)
[0] Intel: Intel(R) Core(TM) i5-6500 CPU @ 3.20GHz, 15985 MB, Max threads(4) 
Time taken for 100 automatic evaluations = 6.77224 secs
Time taken for 100 manual evaluations = 0.968893 secs
```

With https://github.com/arrayfire/arrayfire/pull/1890
```
$ ./examples/helloworld/helloworld_cpu 
ArrayFire v3.5.1 (CPU, 64-bit Linux, build e41c4f7a)
[0] Intel: Intel(R) Core(TM) i5-6500 CPU @ 3.20GHz, 15985 MB, Max threads(4) 
Time taken for 100 automatic evaluations = 3.66398 secs
Time taken for 100 manual evaluations = 0.968872 secs
```

**Comment by pavanky** [comment]
@mchandra More improvements will come through https://github.com/arrayfire/arrayfire/issues/1401

**Comment by pavanky** [comment]
Fixed in hotfixes-3.5.1
